types/vote_test.go

```diff
@@ -336,38 +335,51 @@ func TestVoteString(t *testing.T) {
 	}
 }
 
-func TestVoteValidateBasic(t *testing.T) {
+func TestVoteValidation(t *testing.T) {
 	privVal := NewMockPV()
 
 	testCases := []struct {
-		testName     string
-		malleateVote func(*Vote)
-		expectErr    bool
+		testName                 string
+		malleateVote             func(*Vote)
+		expectBasicErr           bool
+		expectExtErrForPrevote   bool
+		expectExtErrForPrecommit bool
 	}{
-		{"Good Vote", func(v *Vote) {}, false},
-		{"Negative Height", func(v *Vote) { v.Height = -1 }, true},
-		{"Negative Round", func(v *Vote) { v.Round = -1 }, true},
+		{"Good Vote", func(v *Vote) {}, false, false, false},
+		{"Negative Height", func(v *Vote) { v.Height = -1 }, true, true, true},
+		{"Negative Round", func(v *Vote) { v.Round = -1 }, true, true, true},
 		{"Invalid BlockID", func(v *Vote) {
 			v.BlockID = BlockID{[]byte{1, 2, 3}, PartSetHeader{111, []byte("blockparts")}}
-		}, true},
-		{"Invalid Address", func(v *Vote) { v.ValidatorAddress = make([]byte, 1) }, true},
-		{"Invalid ValidatorIndex", func(v *Vote) { v.ValidatorIndex = -1 }, true},
-		{"Invalid Signature", func(v *Vote) { v.Signature = nil }, true},
-		{"Too big Signature", func(v *Vote) { v.Signature = make([]byte, MaxSignatureSize+1) }, true},
+		}, true, true, true},
+		{"Invalid Address", func(v *Vote) { v.ValidatorAddress = make([]byte, 1) }, true, true, true},
+		{"Invalid ValidatorIndex", func(v *Vote) { v.ValidatorIndex = -1 }, true, true, true},
+		{"Invalid Signature", func(v *Vote) { v.Signature = nil }, true, true, true},
+		{"Too big Signature", func(v *Vote) { v.Signature = make([]byte, MaxSignatureSize+1) }, true, true, true},
+		{"Vote extension present", func(v *Vote) { v.Extension = []byte("extension") }, false, true, false},
+		{"Missing vote extension signature", func(v *Vote) { v.ExtensionSignature = nil }, false, false, true},
 	}
 	for _, tc := range testCases {
 		tc := tc
 		t.Run(tc.testName, func(t *testing.T) {
 			ctx, cancel := context.WithCancel(context.Background())
 			defer cancel()
 
-			vote := examplePrecommit(t)
-			v := vote.ToProto()
-			err := privVal.SignVote(ctx, "test_chain_id", v)
-			vote.Signature = v.Signature
-			require.NoError(t, err)
-			tc.malleateVote(vote)
-			assert.Equal(t, tc.expectErr, vote.ValidateBasic() != nil, "Validate Basic had an unexpected result")
+			votes := []*Vote{examplePrevote(t), examplePrecommit(t)}
+			for i, vote := range votes {
```

Comment:
Some of this is pre-existing, but I'd like to put in a plug for getting rid of this loop (instead test each vote separately). I'd also suggest separating these test cases into batches: "ok", "basic validation errors", "prevote errors", and "precommit errors" that do not require flags in the test case.

The fiddly logic to check "is this the specially designated good case" and the various legerdemain with computing Booleans makes it really hard to decipher what is going on when a test fails. The check per vote is short, it's ok to copy paste. It'll all be on the same screenful. 🙂 

It would make a little more code, but IMO tests are a good place to strongly favor legibility and obviousness over compactness IMO.

Schematically:
```go
type testCase struct {
  name         string
  malleateVote func(*Vote)
}

okTests := []testCase{
  // …
}
for _, test := range okTests {
   t.Run(test.name, func(t *testing.T) {
      require.NoError(t, examplePrevote(t).ValidateBasic(), "prevote")
      require.NoError(t, examplePrecommit(t).ValidateBasic(), "precommit")
   })
}
basicErrTests := []testCase{
  // …
}
for _, test := range basicErrTests {
   t.Run(test.anme, func(t *testing.T) {
      require.Error(t, examplePrevote(t).ValidateBasic(), "invalid prevote")
      require.Error(t, examplePrecommit(t).ValidateBasic(), "invalid precommit")      
   })
}

// etc.
```

Reply:
Sounds good :+1: